jupyterhub/templates/scheduling/user-scheduler/deployment.yaml

```diff
@@ -23,6 +23,11 @@ spec:
       {{- if .Values.rbac.enabled }}
       serviceAccountName: user-scheduler
       {{- end }}
+      {{- $_ := merge (dict "podKind" "core") . }}
```

Comment:
These lines ` {{- $_ := merge (dict "podKind" "core") . }}` are I think the hardest to follow, and make contributing to/understanding the chart very difficult. I genuinely don't know what this means or how to use it, so I can't really evaluate whether this is the right thing to do and if/when we might need to do the same or similar thing in other templates.

There's now a lot of hidden logic that is several layers deep.

If, instead, this were a single:

```yaml
{{- include "jupyterhub.tolerations" "core" | nindent 6 }}
```

I feel like that would be simpler and easier to grok as "include the tolerations for a core pod". Is that achievable?

Reply:
Perhaps we can reduce the complexity slightly, I dare not prove a negative, but I have found it very challenging. The merge thing could be simplified by using one or more `set` statements at least.

The reason for the merge, is that you can only pass one object to a helper, and the helper needs a lof of stuff that exists within the default scope referred to as `.` --- By augmenting the default scope, we would change a state globally, which also isnt great. What I have done here is to create a new variable called $_ that creates a new scope by creating a new dictionary, and merging the current scope to it.

https://gitlab.com/charts/gitlab/blob/master/doc/development/README.md#passing-variables-between-control-structures

---

```
{{- include "jupyterhub.tolerations" "core" | nindent 6 }}
```

This helper would not be aware of what `.Release.Name` was if it isnt passed the `.` scope or something that got the `.` scope merged with it. If in the scope of the tolerations helper one would reference `.` it would get `"core"` only. This is sometimes enough for certain helpers, but tend not to be for a lot of helpers.